### PR TITLE
[data] Remove ray.kill in ActorPoolMapOperator

### PR DESCRIPTION
**Pull request by raulchen**
## Why are these changes needed?

With https://github.com/ray-project/ray/pull/47396, now Ray Core can automatically restarts an actor when needing to resubmit tasks. 
This doesn't work with `ray.kill`-ed actors. 
This PR removes ray.kill and let ref counting garbage-collect the actors. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/execution/operators/actor_pool_map_operator.py [hunk 1]
```diff
@@ -640,13 +640,15 @@ def _kill_all_running_actors(self):
 
     def _kill_running_actor(self, actor: ray.actor.ActorHandle):
         """Kill the provided actor and remove it from the pool."""
-        ray.kill(actor)
         del self._num_tasks_in_flight[actor]
+        del self._actor_locations[actor]
```
Comment (bveeramani) [inline python/ray/data/_internal/execution/operators/actor_pool_map_operator.py:644]: Could we add a note that we intend to refactor this with the pause API at some point?

**Review by bveeramani — APPROVED** [review]

File: python/ray/data/_internal/execution/operators/actor_pool_map_operator.py [hunk 2]
```diff
@@ -640,13 +640,15 @@ def _kill_all_running_actors(self):
...
+        assert actor not in self._pending_actors
 
     def _kill_pending_actor(self, ready_ref: ray.ObjectRef):
         """Kill the provided pending actor and remove it from the pool."""
         actor = self._pending_actors.pop(ready_ref)
-        ray.kill(actor)
+        assert actor not in self._num_tasks_in_flight
+        assert actor not in self._actor_locations
```
Comment (alexeykudinkin) [inline python/ray/data/_internal/execution/operators/actor_pool_map_operator.py:651]: Why we're asserting after modified the state?

**Review by alexeykudinkin — COMMENTED** [review]
@raulchen can you please link corresponding ticket

File: python/ray/data/_internal/execution/operators/actor_pool_map_operator.py (hunk 2 above)
Comment (raulchen) [reply python/ray/data/_internal/execution/operators/actor_pool_map_operator.py:651]: the assert is for a different dict. 
it's to make sure the actor handle is not being ref'ed by any of these 3 dicts

**Pull request merged**